unstructured, community, initialize langchain-unstructured package

Coniferish opened the pull request:
#### Update (2): 
A single `UnstructuredLoader` is added to handle both local and api partitioning. This loader also handles single or multiple documents.

#### Changes in `community`:
Changes here do not affect users. In the initial process of using the SDK for the API Loaders, the Loaders in community were refactored. 
Other changes include:
The `UnstructuredBaseLoader` has a new check to see if both `mode="paged"` and `chunking_strategy="by_page"`. It also now has `Element.element_id` added to the `Document.metadata`. `UnstructuredAPIFileLoader` and `UnstructuredAPIFileIOLoader`. As such, now both directly inherit from `UnstructuredBaseLoader` and initialize their `file_path`/`file` attributes respectively and implement their own `_post_process_elements` methods.

--------
#### Update:
New SDK Loaders in a [partner package](https://python.langchain.com/v0.1/docs/contributing/integrations/#partner-package-in-langchain-repo) are introduced to prevent breaking changes for users (see discussion below).

##### TODO:
- [x] Test docstring examples
--------
- **Description:** UnstructuredAPIFileIOLoader and UnstructuredAPIFileLoader calls to the unstructured api are now made using the unstructured-client sdk.
- **New Dependencies:** unstructured-client

- [x] **Add tests and docs**: If you're adding a new integration, please include
  - [x] a test for the integration, preferably unit tests that do not rely on network access,
  - [x] update the description in `docs/docs/integrations/providers/unstructured.mdx`
- [x] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

TODO:
- [x] Update https://python.langchain.com/v0.1/docs/integrations/document_loaders/unstructured_file/#unstructured-api
  - `langchain/docs/docs/integrations/document_loaders/unstructured_file.ipynb`
  - The description here needs to indicate that users should install `unstructured-client` instead of `unstructured`. Read over closely to look for any other changes that need to be made.
- [x] Update the `lazy_load` method in `UnstructuredBaseLoader` to handle json responses from the API instead of just lists of elements.
  - This method may need to be overwritten by the API loaders instead of changing it in the `UnstructuredBaseLoader`.
- [x] Update the documentation links in the class docstrings (the Unstructured documents have moved)
- [x] Update Document.metadata to include `element_id` (see thread [here](https://unstructuredw-kbe4326.slack.com/archives/C044N0YV08G/p1718187499818419))

Coniferish committed:
change inheritance structure for unstructured loaders

Previously, the API Loaders inherited from their respective File/FileIO loaders (which both interhited from the Base Loader). This was problematic for switching from having 'unstructured' as a dependency to 'unstructured-client' because the Base Loader checked if 'unstructured' was installed. Now the 'unstructured'/'unstructured-client' check happens in the get_elements_from_api function and the other Loaders are responsible for checking the packages they depend on. With the API loaders inheriting directly from the Base Loader now, they also now initialize their 'file'/'file_path' attributes and call the super.__init__() method of the Base Loader directly.

Coniferish committed:
fix types

Coniferish committed:
implement using the sdk for making requests to the api

Refactor calls to get_elements_from_api so that Loaders handle single or multiple files. Change metadata_filename variable name in tests/unstructured_kwargs. Import elements_from_json. Etc.

Coniferish committed:
add test, fix default/accepted types (remove None default), remove unused import, move tests to match module

Coniferish committed:
linting

Coniferish commented on libs/community/tests/integration_tests/document_loaders/test_unstructured.py:
```diff
@@ -87,7 +95,7 @@ def test_unstructured_api_file_io_loader() -> None:
             api_key="FAKE_API_KEY",
             strategy="fast",
             mode="elements",
-            file_filename=file_path,
```
The loaders used this `file_filename` parameter I haven't seen before and don't see in the documentation ([here](https://python.langchain.com/v0.2/docs/integrations/providers/unstructured/) or [here](https://api.python.langchain.com/en/latest/document_loaders/langchain_community.document_loaders.unstructured.UnstructuredAPIFileLoader.html)). I assume I need to add logic so this param is still accepted, but please let me know if there's anything else to consider in how that should be handled. I replaced it with `metadata_filename` since that's the param we use in `unstructured`, but that's not actually a recognized [param](https://docs.unstructured.io/api-reference/api-services/api-parameters) for the sdk either.

MthwRobinson replied:
You're good to drop `file_filename`. `file_filename` was deprecated a while back in favor of `metadata_filename`.

https://github.com/Unstructured-IO/unstructured/pull/1191

Coniferish committed:
parameterize test

Coniferish committed:
fix type hint and extract out _get_content helper function

Coniferish committed:
move functions to bottom of the file

Coniferish committed:
refactor _get_content and make file_path required arg in get_elements_from_api

Coniferish committed:
change UnstructuredFileIOLoader's file type hint to include Sequence again and update _get_elements method.

Also add test for UnstructuredFileIOLoader

Coniferish committed:
First pass at implementing UnstructuredAPIFileLoader.lazy_load()

Also, remove API Loaders dependency on unstructured and add comments.

Coniferish commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -45,7 +50,7 @@ def _get_elements(self) -> List:
     @abstractmethod
     def _get_metadata(self) -> dict:
-        """Get metadata."""
+        """Get file_path metadata if available."""
```
The children either return an empty dictionary or `{"source": self.file_path}`.

Coniferish commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -219,6 +224,53 @@ def __init__(
 
         super().__init__(mode=mode, **unstructured_kwargs)
 
+    def lazy_load(self) -> Iterator[Document]:
```
This method was copied from the BaseLoader and updated to expect JSON/a list of dictionaries instead of `Element` objects.

MthwRobinson replied:
Doesn't this method get inherited from the base class? Wondering why you needed to copy it.

Coniferish commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -219,6 +224,53 @@ def __init__(
+                # if hasattr(element, "category"):
+                #     metadata["category"] = element.category
+                yield Document(page_content=element.get("text"), metadata=metadata)
+        elif self.mode == "paged":
```
@MthwRobinson, this `"paged"` mode seems to be equivalent to the `by_page` chunking. I thought about updating the request in the sdk if `self.mode == "paged"`, but I am wondering if there would be any issue with changing this process from post-processing like it is now. Do you think that's safe to do?

MthwRobinson replied:
I would leave this to avoid breaking things for people who are using it. Of course, `chunking_strategy="by_page"` is fine to pass in as well.

Coniferish commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -219,6 +224,53 @@ def __init__(
+                # NOTE(MthwRobinson) - the attribute check is for backward compatibility
+                # with unstructured<0.4.9. The metadata attributed was added in 0.4.9.
+                if element.get("metadata") is not None:
+                    metadata.update(element["metadata"])
```
@MthwRobinson, can you confirm it's safe to remove this check since this implements using `unstructured-client`?

MthwRobinson replied:
Yeah we can remove this check. This was for backward compatibility with a _very very old_ version of `unstructured`. Perhaps add a check on the `unstructured` version with a warning letting people know to upgrade to `unstructured>=0.14.0`.

Coniferish committed:
Make _post_process_elements an @abstractmethod and implement in child classes

Also, change type hints for List/list, Dict/dict, etc. and improve docstrings

Coniferish committed:
move test and fix metadata for api loaders

Coniferish committed:
add lazy_load method to UnstructuredAPIFileIOLoader, split parameterized test into multiple tests, mock calls to the API and add fixtures in tests, fix type hint for get_elements_from_api, etc.

Coniferish committed:
remove type hint for partitioning sequences by UnstructuredFileIOLoader

Coniferish committed:
add  to Document metadata

Coniferish commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -255,44 +212,108 @@ class UnstructuredAPIFileLoader(UnstructuredFileLoader):
-    You can pass in additional unstructured kwargs after mode to apply
-    different unstructured settings.
+    The file loader uses the unstructured partition function and will automatically detect the file
+    type. You can run the loader in one of two modes: "single" and "elements". If you use "single"
```
@MthwRobinson  The Loader also supports a "paged" mode. Should the docstrings be updated to mention this third mode or do we want to continue leaving it out and reference the unstructured's chunking options, which includes chunking `by_page`?

Coniferish committed:
update links and Loader docstrings

Coniferish commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -103,14 +103,14 @@ def lazy_load(self) -> Iterator[Document]:
+    NarrativeText. In addition to these modes (which are specific to the LangChain Loaders)
+    Unstructured has its own "chunking" parameters for post-processing elements into more useful
+    "chunks" for uses cases such as Retrieval Augmented Generation (RAG). You can pass in additional
+    unstructured kwargs to configure different unstructured settings.
```
@MthwRobinson The Loader also supports a "paged" mode. Should the docstrings be updated to mention this third mode or do we want to continue leaving it out and reference the unstructured's chunking options, which includes chunking by_page?

Coniferish replied:
I opted to leave it only mentioning the two modes and reference chunking options

Coniferish replied:
@MthwRobinson, I raised the "paged" mode question here. It seemed like an somewhat irrelevant mode to support here now that we have it via chunking. It wasn't mentioned before and "paged" isn't a functionality added in this PR, so I thought it would be less confusing to tell users two ways to do the same thing. I also did not test what happens if you define both `mode="paged"` and `chunking_strategy="by_page"`

MthwRobinson replied:
Even though paged wasn't added in this PR, the docstrings are out of date and don't include it so might as well clean that up while we're in here.

Since there are current users of the loader who use paged mode, my suggestion would be to emit a warning log letting the user know about `chunking_page="by_page"` and raise a `ValueError` is you set both `chunking_page="by_page"` and `mode="paged"`. I don't think we should remove that mode without having a deprecation warning at least for a while so users have a chance to migrate.

Coniferish committed:
update jupyter notebook and docs

Coniferish committed:
linting and formatting

Coniferish committed:
Merge branch 'master' into jj/sdk

MthwRobinson commented on docs/docs/integrations/providers/unstructured.mdx:
```diff
@@ -8,10 +8,15 @@ ecosystem within LangChain.
 
-- Install the Python SDK with `pip install unstructured`.
+- For the smallest installation footprint and to take advantage of features not available in the
+  open-source `unstructured package`, install the Python SDK with `pip install unstructured-client`.
```
Should the "install `unstructured-client`" part go below in the section you removed about `UnstructuredAPIFileIOLoader`? Don't see the whole file in the diff, but generally:
- For `UnstructuredFileLoader`, tell people to install `unstructured`
- For `UnstructuredAPIFileLoader`, tell people to install `unstructured-client`

Coniferish replied:
Now that the API Loaders don't depend on `unstructured`, I wanted to add an install section for them here. The section below this wasn't completely accurate and seemed redundant once I added this part up here.
You can view the full diff/file [here](https://github.com/langchain-ai/langchain/pull/22779/commits/a172c1fe8aadbf9d1b3e5e6ca96f02dfe8841f2c)

I will update this to specifically mention the loaders, though

MthwRobinson commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -96,18 +66,20 @@ def lazy_load(self) -> Iterator[Document]:
         elif self.mode == "paged":
-            text_dict: Dict[int, str] = {}
-            meta_dict: Dict[int, Dict] = {}
+            text_dict: dict[int, str] = {}
```
Why this change?

Coniferish replied:
This stemmed from the this thread [here](https://unstructuredw-kbe4326.slack.com/archives/C044N0YV08G/p1718187499818419)

MthwRobinson commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -131,14 +103,14 @@ def lazy_load(self) -> Iterator[Document]:
-    You can pass in additional unstructured kwargs after mode to apply
-    different unstructured settings.
+    The file loader uses the unstructured partition function and will automatically detect the file
+    type. You can run the loader in different modes: "single" and "elements". If you use "single"
```
Should we add "paged" into the description here?

MthwRobinson commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -151,20 +123,35 @@ class UnstructuredFileLoader(UnstructuredBaseLoader):
     References
     ----------
-    https://unstructured-io.github.io/unstructured/bricks.html#partition
+    https://docs.unstructured.io/open-source/core-functionality/partitioning
```
Good catch

MthwRobinson commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -151,20 +123,35 @@ class UnstructuredFileLoader(UnstructuredBaseLoader):
         super().__init__(mode=mode, **unstructured_kwargs)
 
-    def _get_elements(self) -> List:
+    def _get_elements(self) -> list:
```
Why this change? Saw a similar one related to `Dict` above.

Coniferish replied:
Just following type-hinting practices I've learned from Steve.

Coniferish replied:
Lmk if you want me to revert these changes or if I can resolve these comments

Coniferish replied:
Actually, here's the PEP for deprecating `typing.List` and other generics. 
https://peps.python.org/pep-0585/#abstract
https://docs.python.org/3/library/typing.html#deprecated-aliases

MthwRobinson commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -181,61 +168,34 @@ def _get_elements(self) -> List:
-    library will split the document into elements such as Title and NarrativeText.
-    You can pass in additional unstructured kwargs after mode to apply
-    different unstructured settings.
+    You can run the loader in one of two modes: "single" and "elements". If you use "single" mode,
```
Same here, should we add "paged" to the description?

MthwRobinson commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -248,51 +208,119 @@ class UnstructuredAPIFileLoader(UnstructuredFileLoader):
         self.url = url
         self.api_key = api_key
 
-        super().__init__(file_path=file_path, mode=mode, **unstructured_kwargs)
```
Why did you need to make this change?

Coniferish replied:
I had to refactor the classes a bit since the API Loaders were not inheriting directly from the BaseUnstructuredLoader and needed to implement some of their methods themselves. This was removed in that process.

MthwRobinson reviewed (changes requested):
Couple of inline comments. Could you also add a blurb to the PR description with instructions on how to test locally?

Coniferish committed:
address unstructured.mdx comment

Coniferish committed:
Address mode='paged'

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish commented:
@ccurme, I believe this is ready for review if you or someone else with LangChain can take a look

MthwRobinson commented on docs/scripts/arxiv_references.py:
```diff
@@ -176,9 +176,9 @@ def search_code_for_arxiv_references(code_dir: Path) -> dict[str, set[str]]:
-            arxiv_id2module_name_and_members_reduced[arxiv_id] = (
-                module_name_and_member_reduced
-            )
+            arxiv_id2module_name_and_members_reduced[
```
Was this just a linting change?

Coniferish replied:
Yes. Came from running `make format` or one of the other commands of theirs

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

efriis reviewed (commented):
Howdy! Could you address merge conflicts and revert changes in unrelated docs files?

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
undo linting changes to unrelated docs

Coniferish committed:
fix notebook merge and mode bug

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish commented:
@efriis, should be good now

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
minor fix

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

efriis reviewed (commented):
I believe this will be a breaking change for people using the existing api connectors because they might not have `unstructured-client` installed.

Would you be interested in doing the api connectors in a partner package, so we can independently version control it to tolerate breaking changes like this?

Coniferish commented:
> I believe this will be a breaking change for people using the existing api connectors because they might not have `unstructured-client` installed.
> 
> Would you be interested in doing the api connectors in a partner package, so we can independently version control it to tolerate breaking changes like this?

That's a good point. @MthwRobinson, what are your thoughts?

Coniferish committed:
linting

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

MthwRobinson commented:
@efriis @Coniferish - Yeah I think that makes sense. That would just live under `libs/partners` and publish as an independent Python package right? If we go that direction, may be good to migrate the other Unstructured loaders there over time as well, though we can certainly start with just the new API loader that uses the client.

@Coniferish - good to go from my end if you want to start working that.

Coniferish committed:
alphabetize loaders

Coniferish committed:
add UnstructuredBaseLoader to package and import tests

Coniferish committed:
init partner package

Coniferish committed:
replicate SDK Loaders in partners

Coniferish committed:
update all references to api url

Coniferish committed:
update docstring and remove unused class

Coniferish committed:
undo changes from rebase attempt

Coniferish committed:
Merge branch 'langchain-ai:master' into jj/sdk

Coniferish committed:
restore API loaders so they don't use the unstrd client, improve test assertions, remove unnecessary splittig of files before partition request, fix SDKLoader Document metadata, etc.

improve test assertions

WIP move sdk tests to new loader test file

remove lazy_load methods from API Loaders since they do not expect json response any more

fix API Loader tests

Change fixtures to return Element objects instead of json and update _post_process_elements methods

refactor to remove unnecessary splitting of files before making partition request

move _get_content tests

rename tests

indicate _get_elements_from_api is private and update tests

fix Document metadata and tests

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish commented:
@MthwRobinson and @efriis, the code has been updated to maintain the API Loaders and add SDK Loaders as partner packages. Do we want to mark the API Loaders as deprecated in favor of the SDK Loaders? Similarly, how do we want to update the documentation?

Coniferish committed:
Merge branch 'master' into jj/sdk

MthwRobinson commented:
@Coniferish - I'm in favor of marking the old API loaders as deprecated in favor of the new ones that use the SDK. I'd recommend logging a warning pointing people toward the new loaders.

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
deprecate API Loaders and update docs to use SDK Loaders

Coniferish commented on libs/partners/unstructured/docs/document_loaders.ipynb:
```diff
@@ -0,0 +1,213 @@
+{
```
This file was auto-generated when initializing the partner package. @efriis, would it be okay to remove this file and just depend on the existing documentation files (`docs/docs/integrations/providers/unstructured.mdx` and `docs/docs/integrations/document_loaders/unstructured_file.ipynb`)? Those have been updated to reference the SDK Loaders instead of the API Loaders.

efriis replied:
yes! And feel free to remove the whole unstructured/docs directory. You have to move these files into the respective directories under root/docs!

Coniferish commented on libs/community/langchain_community/document_loaders/__init__.py:
```diff
@@ -474,6 +474,7 @@
     from langchain_community.document_loaders.unstructured import (
         UnstructuredAPIFileIOLoader,
         UnstructuredAPIFileLoader,
+        UnstructuredBaseLoader,
```
`UnstructuredBaseLoader` was added to this and other files like it to make it accessible for import as a parent class for the SDK Loaders.

Coniferish replied:
Good call. I'll prefix it.

Coniferish replied:
@efriis, The tests are failing because I changed this to `_UnstructuredBaseLoader` and that class is used in places I didn't anticipate--it's used in `libs/community/langchain_community/document_loaders/lakefs.py` and `libs/community/langchain_community/document_loaders/s3_file.py` for example. I'm not familiar with these loaders, but should I revert making the BaseLoader private? Happy to undo that or fix all the references so they refer to `_UnstructuredBaseLoader`. Lmk

Coniferish committed:
fix docstring

Coniferish commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -183,60 +195,40 @@ def _get_elements(self) -> List:
+    since="0.2.8",
+    removal="0.4.0",  # TODO (jennings): determine when to set removal
+    alternative_import="langchain_unstructured.UnstructuredSDKFileLoader",
+)
```
@MthwRobinson, is version 0.4.0 a reasonable benchmark for removing these loaders?

MthwRobinson replied:
Yeah that seems like plenty of warning!

Coniferish committed:
Merge branch 'master' into jj/sdk

MthwRobinson commented on libs/partners/unstructured/langchain_unstructured/document_loaders.py:
```diff
@@ -0,0 +1,369 @@
+        elements_json = self._get_elements()
+        self._post_process_elements(elements_json)
+
+        if self.mode == "elements":
```
Yeah since we're moving to a new loader class here, I think it would make sense to make those updates.

Coniferish committed:
update test assertions and remove 'mode' param from SDK loaders

Coniferish committed:
add tests for mode

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
remove comments

efriis commented:
Feel free to mark ready to review when it's good for another review/merge!

Coniferish commented on docs/docs/integrations/providers/unstructured.mdx:
```diff
@@ -8,10 +8,17 @@ ecosystem within LangChain.
 
-- Install the Python SDK with `pip install unstructured`.
+- For the smallest installation footprint and to take advantage of features not available in the
+  open-source `unstructured` package, install the Python SDK with `pip install unstructured-client`
```
Ah, both actually. `unstructured-client` is a dependency for `langchain-unstructured`. I'll update that and the instructions below that just reference `unstructured`

Coniferish commented on docs/docs/integrations/providers/unstructured.mdx:
```diff
@@ -23,15 +30,7 @@ its dependencies running locally.
     - `libreoffice` (MS Office docs)
     - `pandoc` (EPUBs)
 
-When running locally, Unstructured also recommends using Docker [by following this guide](https://docs.unstructured.io/open-source/installation/docker-installation)
```
Good catch. I think I missed that when resolving conflicts when rebasing

Coniferish committed:
address comments about private class, documentation, etc.

Coniferish committed:
remove libs/partners/unstructured/docs/document_loaders.ipynb

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
linting

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
Merge branch 'master' into jj/sdk

efriis commented:
![Screenshot 2024-07-15 at 1 52 27 PM](https://github.com/user-attachments/assets/a7363462-689c-454b-bc79-d7641eaa771d)
error from docs build

efriis commented:
And I believe the other CI job logs should be viewable in the "details" links above!

Coniferish committed:
poetry lock --no-update

Coniferish committed:
change all references back to UnstructuredBaseLoader

Coniferish committed:
Implement UnstructuredLoader and update unit tests after making methods lazy properties

Coniferish committed:
update tests

Coniferish committed:
wip

Coniferish committed:
refactor to simplify interface and misc.

Coniferish committed:
fix classes to pass tests

Coniferish committed:
update docs and make file_path a positional arg

Coniferish commented on docs/docs/integrations/providers/unstructured.mdx:
```diff
@@ -42,30 +47,21 @@ Check out the instructions
 
 `CHM` means `Microsoft Compiled HTML Help`.
 
-See a usage example in the API documentation.
```
I did not see an exiting example in the API documentation, so I removed this to not mislead users.

Coniferish committed:
remove UnstructuredBaseLoader as a public class and references to SDK loaders

Coniferish committed:
add deprecation decorators, undo some refactoring, and add type hints to libs/community/langchain_community/document_loaders/unstructured.py

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
update README

Coniferish committed:
linting and type hinting

Coniferish committed:
add SDK example to docs

Coniferish committed:
Merge branch 'master' into jj/sdk

Coniferish committed:
add unstructured to list of providers

efriis commented on libs/langchain/langchain/document_loaders/unstructured.py:
revert

efriis commented on poetry.lock:
revert

efriis commented on pyproject.toml:
revert

efriis commented on libs/partners/unstructured/pyproject.toml:
```diff
@@ -0,0 +1,87 @@
+license = "MIT"
+
+[tool.poetry.urls]
+"Source Code" = "https://github.com/langchain-ai/langchain/tree/master/libs/partners/unstructured"
```
Erick todo: add release notes link

efriis commented on libs/partners/unstructured/pyproject.toml:
```diff
@@ -0,0 +1,87 @@
+
+[tool.poetry.dependencies]
+python = ">=3.8.1,<4.0"
+langchain-core = "^0.2.0"
```
make this more recent

Coniferish replied:
This file was generated when I initialized the partner package with `langchain-cli integration new` I believe. I don't have much experience with pyproject.toml files, but are you saying I should update this part manually?

efriis commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -1,14 +1,22 @@
 from langchain_community.document_loaders.base import BaseLoader
 
+if TYPE_CHECKING:
+    from unstructured.documents.elements import Element
```
i believe this will fail ci - you may have to type these as `Any` in community

efriis commented on libs/community/langchain_community/document_loaders/unstructured.py:
```diff
@@ -41,44 +49,54 @@ class UnstructuredBaseLoader(BaseLoader, ABC):
 
     @abstractmethod
-    def _get_elements(self) -> List:
+    def _get_elements(self) -> list[Element]:
```
revert - leave as `List` to avoid ci lint issues

efriis commented on libs/community/langchain_community/document_loaders/unstructured.py:
could you summarize the changes in community? Are any of these changes breaking for existing users?

Coniferish replied:
👍 Added to the PR description

Coniferish committed:
address comments

Coniferish committed:
Merge branch 'langchain-ai:master' into jj/sdk

Coniferish committed:
revert files

Coniferish committed:
refactor to simplify diff

Coniferish committed:
Merge branch 'master' into jj/sdk

Commit:
Merge branch 'master' into jj/sdk

Commit:
x

Coniferish committed:
add return values and address CI errors

Coniferish committed:
Merge branch 'master' into jj/sdk

Commit:
format

Commit:
Merge branch 'master' into jj/sdk

isahers1 committed:
docs: add tables for search and code interpreter tools (#24586)

Commit:
x

Commit:
x

Commit:
cli: remove snapshot flag from pytest defaults (#24622)

Commit:
milvus: release 0.1.3 (#24624)

zc277584121 committed:
partners[milvus]: add dynamic field (#24544)

add dynamic field feature to langchain_milvus
more unittest, more robustic

plan to deprecate the `metadata_field` in the future, because it's
function is the same as `enable_dynamic_field`, but the latter one is a
more advanced concept in milvus

Signed-off-by: ChengZi <chen.zhang@zilliz.com>
Co-authored-by: Erick Friis <erick@langchain.dev>

Commit:
cli: release 0.0.26 (#24623)

- **cli: remove snapshot flag from pytest defaults**
- **x**
- **x**

Coniferish committed:
change client to UnstructuredClient, add os.getenv(), and update jupyter notebook

Coniferish committed:
linting

Coniferish committed:
fix TypeAlias import

Commit:
Merge branch 'master' into jj/sdk

Commit:
x

Coniferish committed:
linting

Commit:
x

Commit:
Merge branch 'jj/sdk' of github.com:Coniferish/langchain into jj/sdk

Commit:
Merge branch 'master' into jj/sdk

Commit:
x

Commit:
x

Commit:
x

Commit:
x

Commit:
x

Commit:
x
